src/Sentry/Internal/Http/CachingTransport.cs

```diff
@@ -0,0 +1,266 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Sentry.Extensibility;
+using Sentry.Protocol.Envelopes;
+
+namespace Sentry.Internal.Http
+{
+    internal class CachingTransport : ITransport, IDisposable
+    {
+        private const string EnvelopeFileExt = "envelope";
+
+        private readonly ITransport _innerTransport;
+        private readonly SentryOptions _options;
+        private readonly string _cacheDirectoryPath;
+
+        // When a file is getting processed, it's moved to a child directory
+        // to avoid getting picked up by other threads.
+        private readonly string _processingDirectoryPath;
+
+        // Signal that tells the worker whether there's work it can do.
+        // Pre-released because the directory might already have files from previous sessions.
+        private readonly Signal _workerSignal = new Signal(true);
+
+        // Lock to synchronize file system operations inside the cache directory.
+        // It's required because there are multiple threads that may attempt to both read
+        // and write from/to the cache directory.
+        // Lock usage is minimized by moving files that are being processed to a special directory
+        // where collisions are not expected.
+        private readonly Lock _cacheDirectoryLock = new Lock();
+
+        private readonly CancellationTokenSource _workerCts = new CancellationTokenSource();
+        private readonly Task _worker;
+
+        public CachingTransport(ITransport innerTransport, SentryOptions options)
+        {
+            _innerTransport = innerTransport;
+            _options = options;
+
+            _cacheDirectoryPath = !string.IsNullOrWhiteSpace(options.CacheDirectoryPath)
+                ? _cacheDirectoryPath = options.CacheDirectoryPath
+                : throw new InvalidOperationException("Cache directory is not set.");
+
+            _processingDirectoryPath = Path.Combine(_cacheDirectoryPath, "__processing");
+
+            // Processing directory may already contain some files left from previous session
+            // if the worker has been terminated unexpectedly.
+            // Move everything from that directory back to cache directory.
+            if (Directory.Exists(_processingDirectoryPath))
+            {
+                foreach (var filePath in Directory.EnumerateFiles(_processingDirectoryPath))
+                {
+                    Directory.CreateDirectory(_cacheDirectoryPath);
```

Comment:
This could be created outside of the `foreach`.
It should no-op if it already exists.